geopmdpy/geopmdpy/stats.py

```diff
@@ -0,0 +1,152 @@
+        """Create stats collector
+
+        Provide a list of read requests for PlatformIO.
+
+        Args:
+            signal_config (list((str, int, int))): List of requested
+                signals where each tuple represents
+                (signal_name, domain_type, domain_idx).
+
+        """
+        global _dl
+        self._collector_ptr = None
+        signal_config = [(rr[0], rr[1], rr[2]) for rr in signal_config]
+        num_signal = len(signal_config)
+        if num_signal == 0:
+            raise RuntimeError('Collector creation failed: length of input is zero')
+
+        signal_config_carr = gffi.gffi.new(f'struct geopm_request_s[{num_signal}]')
+        for idx, req in enumerate(signal_config):
+            signal_config_carr[idx].domain = req[1]
+            signal_config_carr[idx].domain_idx = req[2]
+            signal_config_carr[idx].name = req[0].encode()
+
+        collector_ptr = gffi.gffi.new('struct geopm_stats_collector_s **');
+
+        err = _dl.geopm_stats_collector_create(num_signal, signal_config_carr, collector_ptr)
+        if err < 0:
+            raise RuntimeError('geopm_stats_collector() failed: {}'.format(error.message(err)))
+        self._collector_ptr = collector_ptr[0];
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.close()
+
+    def close(self):
+        """Free all resources by deleting the StatsCollector
+
+        The object cannot be used after this call.  Must be called on
+        all Collector objects created without context management when
+        they are no longer in use.
+
+        """
+        global _dl
+        if (self._collector_ptr is not None):
+            collector_ptr = self._collector_ptr
+            self._collector_ptr = None
+            _dl.geopm_stats_collector_free(collector_ptr)
+
+    def _check_ptr(self, func_name):
+        if self._collector_ptr is None:
+            raise RuntimeError(f'Called Collector.{func_name}() after calling Collector.close()')
+
+    def update(self):
+        """Update the collector with new values
+
+        User is expected to call geopmdpy.pio.read_batch() prior to calling this
+        interface.  The sampled values will be used to update the report statistics.
+
+        """
+        global _dl
+        self._check_ptr('update')
+        err = _dl.geopm_stats_collector_update(self._collector_ptr)
+        if err < 0:
+            raise RuntimeError('geopm_stats_collector_update() failed: {}'.format(error.message(err)))
+
+    def report_yaml(self):
+        """Create a yaml report
+
+        Return a yaml string containing a report that shows all statistics
+        gathered by calls to update().
+
+        """
+        global _dl
+        self._check_ptr('report_yaml')
+        report_max = gffi.gffi.new("size_t *")
+        report_max[0] = 0
+        _dl.geopm_stats_collector_report_yaml(self._collector_ptr, report_max, gffi.gffi.NULL)
```

Comment:
If we start doing scenarios with multiple reports generated per session, then we might consider caching the expected `report_max` inside the Collector instance so we don't need to run the yaml-generating code twice per report.

I think the current implementation is good for our current single-report-per-session case. Maybe just add a note to look into this in the issue tracker for the multi-report-per-session work.

Reply:
The issue with doing that in the C interface is the const-ness of the collector pointer in the interface.  We could cache in python however.

Reply:
We could also mark the cache as a `mutable` member

Reply:
https://github.com/geopm/geopm/issues/3553